src/Symfony/Component/String/AbstractUnicodeString.php

```diff
@@ -450,6 +474,18 @@ public function upper(): static
         return $str;
     }
 
+    public function localeUpper(string $locale): static
+    {
+        if (null !== $transliterator = $this->getLocaleTransliterator($locale, 'Upper')) {
```

Comment:
Yes, at the moment it only affects "TR", "AZ", "EL" and "NL" and these are all used in the unittest's dataprovider + I added "EN"/"EN_US" to show the difference.
Are there more testcases you want me to add?

Reply:
Ah I understand, I thought you meant a PHPUnit test for the hot path.
I'd vote against having a hardcoded set, sure now it's only TR/AZ/EL/NL, but we don't know if that stays the same in the future, and first checking Transliterator::listIDs, would break fetching NL-Title when passing nl_NL as locale.

btw, Transliterator::listIDs contains "el-Title", not "el_GR-Title", but when passing "el_GR-Title" it does return an instance.. However this doesn't work for "nl_NL-Title", that has to be "nl-Title".

Do you have an idea on how to pick this up?

Reply:
> Transliterator::listIDs contains "el-Title", not "el_GR-Title", but when passing "el_GR-Title" it does return an instance.. However this doesn't work for "nl_NL-Title", that has to be "nl-Title".

Well that could be something to specify in the documentation :) 

> sure now it's only TR/AZ/EL/NL, but we don't know if that stays the same in the future

That's why i was thinking at first to parse the offocial ICU docs (not on real time of course, once in a while to update the script)... but that is a story for another time, i do not want to slow you down (even if it seems so)

;)

Reply:
With the current implementation of looking at input and also parent locale, it always works.
I'll add a comment to the new localeX methods, to describe the expected input

Reply:
There's one reason that could make @smnandre's proposal worth it: instantiation time. IIRC, transliterators can be slow to create. Can you have a look at this from this angle after doing some benchs please?

Reply:
Quick benchmark
```php
$string = static::createFromString('istanbul');
$iterations = 1_000_000;

$startTime = microtime(true);
$i = $iterations;
while (--$i) {
    $string->upper();
}
$endTime = microtime(true);
echo $endTime - $startTime, PHP_EOL;

$startTime = microtime(true);
$i = $iterations;
while (--$i) {
    $string->localeUpper('tr_TR');
}
$endTime = microtime(true);
echo $endTime - $startTime, PHP_EOL;

$startTime = microtime(true);
$i = $iterations;
while (--$i) {
    $string->localeUpper('en_US');
}
$endTime = microtime(true);
echo $endTime - $startTime, PHP_EOL;
```

Results in:
1.7240459918976
3.6631438732147
3.0851650238037

And if I place `return null;` on top of the getLocaleTransliterator method, I can still get:
1.5016701221466
2.4766409397125
2.4534759521484
as times.

Reply:
This is something we can optimize in a following step.. but indeed there is a big memory difference :| 

I simplified the two options 

# Option One

```php
<?php

$word = 'étienne';
$locale = 'fr';

if (preg_match('/^(EL|AB|CD|EF|GH|IJ)/i', $locale)) {
    $transliterator = \Transliterator::create($locale.'-UPPER');
    echo 'A' . ($word = $transliterator->transliterate($word));
    return $word;
}

echo 'B' . ($word = mb_strtoupper($word));
return $word;
```

https://3v4l.org/FArNT/perf


### Option Two

```php
<?php

$word = 'étienne';
$locale = 'fr';

if (null !== $transliterator = Transliterator::create($locale.'-UPPER')) {
    echo 'A' . ($word = $transliterator->transliterate($word));
    return $word;
}

echo 'B' . ($word = mb_strtoupper($word));
return $word;
```

https://3v4l.org/NsKEe/perf

## Results

### One - "fr"

<img width="1015" alt="Capture d’écran 2023-10-23 à 14 44 07" src="https://github.com/symfony/symfony/assets/1359581/3c863a14-3a4b-4137-b5be-5c03dff319ba"> 

### One - "el" 

<img width="1030" alt="Capture d’écran 2023-10-23 à 14 44 19" src="https://github.com/symfony/symfony/assets/1359581/16a93945-a2a4-4bb4-8478-7e51b711db66"> 

### Two - "fr" 

<img width="1012" alt="Capture d’écran 2023-10-23 à 14 44 27" src="https://github.com/symfony/symfony/assets/1359581/4060150e-ca74-43a9-a464-20f2538cf9e5">

### Two - "el" 

<img width="1009" alt="Capture d’écran 2023-10-23 à 14 44 42" src="https://github.com/symfony/symfony/assets/1359581/86ee8d81-731a-4023-b527-9a7410505eac">


I'm sincerly surprised, but there is a **massive difference** for the "FR" path (any langage with no transliterator available)


I do not have the time right now, but we should check with the transliterator_ids method

Reply:
Sure! Note that this should be merged in 7.1 since 6.4/7.0 are closed now.

Reply:
Just missed the window? That's too bad.
I don't see a 7.1 branch to base this PR on, what should I do?

Reply:
Just be patient - we'll resume merges for 7.1 at the end of November?

Reply:
Alright, sorry. Thanks for the help.